File: solutions/Leetcode_554/leetcode_554.py
from typing import List


class Solution1:
    def least_bricks(self, wall: List[List[int]]) -> int:
        edge_dict = {0: 0}
        for row in wall:
            position = 0
            for x in row[:-1]:
                position += x
                edge_dict[position] = edge_dict.get(position, 0) + 1
        return len(wall) - max(edge_dict.values())


if __name__ == '__main__':
    wall = [[1, 2, 2, 1],
            [3, 1, 2],
            [1, 3, 2],
            [2, 4],
            [3, 1, 2],
            [1, 3, 1, 1]]
    ans = Solution1().least_bricks(wall)
    print(ans)

File: solutions/Leetcode_554/leetcode_554.cc
#include <map>
#include <vector>

using namespace std;

class Solution1
{
public:
    int leastBricks(vector<vector<int>> &wall)
    {
        map<int, int> edge_dict = {{0, 0}};
        int max_count = 0;
        for (vector<int> &row : wall)
        {
            int position = 0;
            for (int i = 0; i != row.size() - 1; ++i)
            {
                position += row[i];
                edge_dict[position]++;
                max_count = max(max_count, edge_dict[position]);
            }
        }
        return wall.size() - max_count;
    }
};